include/yaml-cpp/node/convert.h

```diff
@@ -161,109 +170,222 @@ struct convert<bool> {
   YAML_CPP_API static bool decode(const Node& node, bool& rhs);
 };
 
-// std::map
-template <typename K, typename V>
-struct convert<std::map<K, V>> {
-  static Node encode(const std::map<K, V>& rhs) {
+// Helper class for converting something which is "sequence-like" to and from a
+// Node.
+template <class T>
+struct convert_as_sequence {
+  static Node encode(const T& sequence) {
+    Node node(NodeType::Sequence);
+    for (const auto& item : sequence)
+      node.push_back(item);
+    return node;
+  }
+  static bool decode(const Node& node, T& sequence) {
+    if (!node.IsSequence())
+      return false;
+    sequence.clear();
+    for (const auto& item : node)
+      sequence.push_back(item.template as<typename T::value_type>());
+    return true;
+  }
+};
+
+// Helper class for converting something which is "map-like" to and from a Node.
+template <class T, class Key = typename T::key_type,
+          class Value = typename T::mapped_type>
+struct convert_as_map {
+  static Node encode(const T& map) {
     Node node(NodeType::Map);
-    for (typename std::map<K, V>::const_iterator it = rhs.begin();
-         it != rhs.end(); ++it)
-      node.force_insert(it->first, it->second);
+    for (const auto& kv : map)
+      node.force_insert(kv.first, kv.second);
     return node;
   }
 
-  static bool decode(const Node& node, std::map<K, V>& rhs) {
+  static bool decode(const Node& node, T& map) {
     if (!node.IsMap())
       return false;
+    map.clear();
+    for (const auto& kv : node) {
+// if we're dealing with GCC (note: clang also defines __GNUC__)
+#if defined(__GNUC__)
+#if __GNUC__ < 4
+      // gcc version < 4
+      map.insert(std::make_pair(kv.first.template as<Key>(),
+                                kv.second.template as<Value>()));
+#elif __GNUC__ == 4 && __GNUC_MINOR__ < 8
+      // 4.0 <= gcc version < 4.8
+      map.insert(std::make_pair(kv.first.as<Key>(), kv.second.as<Value>()));
+#else
+      // 4.8 <= gcc version
+      map.emplace(kv.first.as<Key>(), kv.second.as<Value>());
+#endif  // __GNUC__ < 4
+#else
+      // for anything not GCC or clang...
+      // probably some more #ifdef guards are needed for MSVC.
+      map.emplace(kv.first.as<Key>(), kv.second.as<Value>());
+#endif  // defined(__GNUC__)
+    }
+    return true;
+  }
+};
+
+// Helper class for converting something which is "set-like" to and from a Node.
+// A set is realized as a yaml sequence.
+template <class T>
+struct convert_as_set {
```

Comment:
I just checked what PyYAML does and it follows the spec literally:

```python
>>> import yaml
>>> aset = {2,3,5,7,11,13,13,13,13,13,13}
>>> yaml.dump(aset)
'!!set {2: null, 3: null, 5: null, 7: null, 11: null, 13: null}\n'
```
I don't like this either because:
1. It makes for an awkward syntax with the "?" symbol when you have complex keys
2. Redundant ": null" substrings

It's your choice though, let me know what you want so I can adjust the converters and tests accordingly. Me personally I vote for sets realized as sequences to maintain human readability.